move hostap_iface from beerocks_agent.conf to prplmesh

adam1985d opened the pull request:
this PR should follow with similar PR to prplmeshOpenWrt feed that adds
the interfaces to the prplmesh config file.
#787

vitalybu commented on agent/config/beerocks_agent.conf.in:
```diff
@@ -29,20 +29,17 @@ backhaul_wire_iface=@BEEROCKS_BH_WIRE_IFACE@
 [agent0]
 radio_identifier=00:00:00:00:00:00
 hostap_iface_type=WIFI_INTEL
-hostap_iface=@BEEROCKS_WLAN1_IFACE@
 hostap_ant_gain=0
 enable_repeater_mode=@BEEROCKS_REPEATER_MODE@
```
Any reason to not move everything into the UCI?
@tomereli, thoughts?

adam1985d replied:
In the task description, it is mentioned that all information is planned to be moved to the UCI.

I've decided to split it to PRs since this PR is a prerequisite for UGW prplmesh operational.

after we complete the UGW fix, we can continue the porting of the rest of the params.

tomereli replied:
We can move everything to UCI but in appropriate hierarchies (not flat).
But I agree with Adam, not part of this PR.

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -299,13 +306,14 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
-            slave_ap_ifaces.insert(beerocks_slave_conf.hostap_iface[slave_num]);
+    for (auto &elm : interfaces_map) {
+        if (!elm.second.empty()) {
+            slave_ap_ifaces.insert(elm.second);
```
Why do we need another data structure?
If we already create a map with the interfaces (and potentially all other related data), why not use everywhere?

adam1985d replied:
the BH manager needs a list of all interfaces, while the interfaces_map maps agent/radio index to hostap_iface.
we can provide the map to the backhaul and let it take care of extracting the information but I think that changing the CTOR signature will hide the actual "requirements" of the BH manager.

I leave the call to you.

tomereli replied:
One map to rule them all!
Interface map is basically a radio map, which should have the VAPs and all other related data as Vitaly suggests. But since it is a bigger change I'm not sure.. @vitalybu ?

adam1985d replied:
I can change the interface_map to be radio_map and expand it to contain the rest of the param values as part of the PR that moves the rest of the params to the prplmesh config file.
Unless you think it must be part of this PR, and I will need to change the backhaul_manager implementation accordingly.

adam1985d replied:
talked with @vitalybu and the suggested change will be handled as part of the PR to move rest of params to the prplmesh config file

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -388,10 +396,9 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
 
 static int run_son_slave(int slave_num, beerocks::config_file::sConfigSlave &beerocks_slave_conf,
-                         int argc, char *argv[])
+                         const std::string hostap_iface, int argc, char *argv[])
```
```suggestion
                         const std::string& hostap_iface, int argc, char *argv[])
```

adam1985d replied:
fixed

vitalybu commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -232,5 +232,52 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
+        return RETURN_ERR;
+    }
+
+    std::string iface_str("wlan" + std::to_string(radio_num));
```
Maybe the interface prefix should not be hard-coded.
Can be moved to agent config file or the UCI DB.
@tomereli?

vitalybu replied:
Also, where do you actually look for the UCI key?

tomereli replied:
This is indeed wrong and beats the purpose - we don't want any dependency on the interface name. It should be set in /etc/config/prplmesh once and that's it - no radio_num dependency should be there.

adam1985d replied:
this is Linux BPL - so no /etc/config/prplmesh
in order to read this configuration from the prplmesh_platform_db file, I need to add infrastructure for reading non-flat radio configuration from the prplmesh_platform_db file.

do we want this to be part of this PR?

adam1985d replied:
talked with @tomereli and since this is the Linux BPL - we decided to leave it as is.

vitalybu commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -232,5 +232,22 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
 
+int cfg_get_hostap_iface(int32_t radio_num, char hostap_iface[BPL_IFNAME_LEN])
+{
+    if (!hostap_iface) {
+        MAPF_ERR("cfg_get_hostap_iface: invalid input: hostap_iface is NULL");
+        return RETURN_ERR;
+    }
+
+    if (radio_num < 0) {
+        MAPF_ERR("cfg_get_hostap_iface: invalid input: radio_num < 0");
+        return RETURN_ERR;
+    }
```
I have a personal preference of validating the arguments in the same order as they are defined in the function (nitpick)

adam1985d replied:
me too - but I prefer to check invalid pointer before invalid range.

vitalybu commented on framework/platform/bpl/uci/cfg/bpl_cfg.cpp:
```diff
@@ -276,19 +276,34 @@ int cfg_get_administrator_credentials(char pass[BPL_PASS_LEN]) { return 0; }
 
+int cfg_get_hostap_iface(int32_t radio_num, char hostap_iface[BPL_IFNAME_LEN])
+{
+    if (!hostap_iface) {
+        MAPF_ERR("cfg_get_hostap_iface: invalid input: hostap_iface is NULL");
+        return RETURN_ERR;
+    }
+
+    if (radio_num < 0) {
+        MAPF_ERR("cfg_get_hostap_iface: invalid input: radio_num < 0");
+        return RETURN_ERR;
+    }
```
I have a personal preference of validating the arguments in the same order as they are defined in the function (nitpick)

adam1985d replied:
me too, but I prefer to check invalid pointers before an invalid range.

vitalybu commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -249,5 +249,35 @@ int cfg_get_hostap_iface(int32_t radio_num, char hostap_iface[BPL_IFNAME_LEN])
     return RETURN_OK;
 }
 
+int cfg_get_all_prplmesh_wifi_interfaces(BPL_WLAN_IFACE *interfaces, uint8_t *num_of_interfaces)
```
Maybe better to change the `num_of_interfaces` to `int` to prevent casting.

adam1985d replied:
I changed `num_of_interfaces` to `int`

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -145,7 +145,8 @@ static std::string get_sta_iface_from_hostap_iface(const std::string &hostap_ifa
 static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_slave_conf,
-                                  son::slave_thread::sSlaveConfig &son_slave_conf, int slave_num)
+                                  son::slave_thread::sSlaveConfig &son_slave_conf,
+                                  const std::string hostap_iface, int slave_num)
```
```suggestion
                                  const std::string &hostap_iface, int slave_num)
```

adam1985d replied:
added the `&` as suggested. thanks

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -498,6 +502,25 @@ int main(int argc, char *argv[])
+
+    //create unordered_map of interfaces
+    std::unordered_map<int, std::string> interfaces_map;
+    for (size_t i = 0; i < num_of_interfaces; i++) {
```
Why `size_t` and not `int`?

adam1985d replied:
changed to`int` as part of change of num_of_interfaces to `int`

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -186,21 +186,25 @@ static void son_slave_watchdog(beerocks::config_file::sConfigSlave &beerocks_sla
-                {
-                    file_name = BEEROCKS_BIN_PATH + std::string(BEEROCKS_AGENT);
+        auto hostap_iface_elm = interfaces_map.find(slave_num);
+        if (hostap_iface_elm != interfaces_map.end()) {
```
Why not return in case the interface is not found and prevent one level of indentation?
```suggestion
        if (hostap_iface_elm == interfaces_map.end()) {
            return;
        }
        ...
```

adam1985d replied:
the suggestion is taken. changed to a negative check. if the negative condition is met - continue.

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -186,21 +186,25 @@ static void son_slave_watchdog(beerocks::config_file::sConfigSlave &beerocks_sla
+        auto hostap_iface_elm = interfaces_map.find(slave_num);
+        if (hostap_iface_elm != interfaces_map.end()) {
+            std::string hostap_iface = hostap_iface_elm->second;
+            if (!hostap_iface.empty()) {
```
Same here. Prevent another level of indentation.
```suggestion
            if (hostap_iface.empty()) {
                return;
            }
            ...
```

adam1985d replied:
the suggestion is taken. changed to a negative check. if the negative condition is met - continue.

vitalybu commented on framework/platform/bpl/uci/cfg/bpl_cfg.cpp:
```diff
@@ -324,8 +324,8 @@ int cfg_get_all_prplmesh_wifi_interfaces(BPL_WLAN_IFACE *interfaces, uint8_t *nu
-            MAPF_ERR("cfg_get_all_prplmesh_wifi_interfaces: failed to get wifi interface for radio"
-                     << index);
+            MAPF_DBG("cfg_get_all_prplmesh_wifi_interfaces: failed to get wifi interface for radio"
+                     << index << "or radio" << interfaces_count << ".hostap_iface doesn't exist");
```
```suggestion
                     << index << " or radio" << interfaces_count << ".hostap_iface doesn't exist");
```

adam1985d replied:
fixed

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -222,7 +222,7 @@ int cfg_get_administrator_credentials(char pass[BPL_PASS_LEN]) { return RETURN_E
 int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAME_LEN])
 {
-    if (iface == NULL || sta_iface == NULL) {
+    if (!iface || !sta_iface) {
```
Please mention this in the commit message:

> In addition, fix some KW optential issues with NULL comparison

adam1985d replied:
added to the commit message

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -232,5 +232,22 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
+        return RETURN_ERR;
+    }
+
+    if (radio_num < 0) {
```
So why not declare radio num as unsigned?

adam1985d replied:
talked with @tomereli offline and we can skip this comment

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -232,5 +232,22 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
+        return RETURN_ERR;
+    }
+
+    std::string iface_str("wlan" + std::to_string(radio_num * 2));
```
Please add a comment explaining what we expect in the platform_db in Linux mode.
Also, I expect to see such a change in prplmesh_utils.sh which creats the database.

adam1985d replied:
added the following comment:
```
// the linux implementation expects to receive "wlanX" for interface names where the X is an
// even index starting with 0 (e.g.: radio0:wlan0, radio1:wlan2, radio2:wlan4 etc.)
```

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -467,6 +467,17 @@ int cfg_get_administrator_credentials(char pass[BPL_USER_PASS_LEN]);
  */
 int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAME_LEN]);
 
+/**
```
Your change is not only for UCI.
The headline is not according to the format we use.

Suggested commit message:

> bpl: add cfg_get_hostap_iface API
>
> Preparative commit.
Add infrastructure to get hostap_iface param from platform DB. 
The hostap_iface is available per radio agent (son slave) and if it is set,
a process (`beerocks_agent -i<iface>`) for that interface is created.
This is instead of the current state which gets the interface name from the config file `beerocks_agent.conf`.
>
> Signed-off-by...

adam1985d replied:
changed the commit message.

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -225,6 +225,12 @@ struct BPL_WLAN_PARAMS {
     char security[BPL_SEC_LEN];
 };
 
+struct BPL_WLAN_IFACE {
```
Please add doxygen for the new struct.

adam1985d replied:
added the doxygen

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -225,6 +225,12 @@ struct BPL_WLAN_PARAMS {
 
+struct BPL_WLAN_IFACE {
+    int radio_num;
+    /* Wi-Fi interface name */
```
Self-explanatory (nitpick)

adam1985d replied:
removed the comment

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -478,6 +484,17 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
 int cfg_get_hostap_iface(int32_t radio_num, char hostap_iface[BPL_IFNAME_LEN]);
 
+/**
+ * Returns all the configured HOSTAP interfaces
```
What does it mean "configured HOSTAP interfaces"? Please add an explanation.

adam1985d replied:
does the following headline provide enough detail or do you think the additional information is required?
```
Returns all the HOSTAP interfaces available in prplmesh config file
```

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg.cpp:
```diff
@@ -305,5 +305,36 @@ int cfg_get_hostap_iface(int32_t radio_num, char hostap_iface[BPL_IFNAME_LEN])
+        MAPF_ERR("cfg_get_all_prplmesh_wifi_interfaces: invalid input: num_of_interfaces is NULL");
+        return RETURN_ERR;
+    }
+    if (*num_of_interfaces < 1) {
```
I prefer to compare to 0 (nitpick)

adam1985d replied:
in response to a previous comment, the type was changed to int, thus this condition remains as is.

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -225,6 +225,12 @@ struct BPL_WLAN_PARAMS {
     char security[BPL_SEC_LEN];
 };
```
Commit message headline should be "bpl: add cfg_get_all_prplmesh_wifi_interfaces API"
What is "prpl" interfaces"? don't use this terminology.

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -145,7 +145,8 @@ static std::string get_sta_iface_from_hostap_iface(const std::string &hostap_ifa
 
 static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_slave_conf,
-                                  son::slave_thread::sSlaveConfig &son_slave_conf, int slave_num)
+                                  son::slave_thread::sSlaveConfig &son_slave_conf,
```
Commit message headline should be "agent: get hostap_iface from platform DB"

Commit message suggestion:

>agent: get hostap_iface from platform DB
>
> Preparative commit.
>This commit utilizes the new API to read all the hostap interfaces
from the platform database instead of the agent config file, so that each platform can have
its own different interface names.
Interfaces are read into an unordered map of radio index to interface name which is propagated to all of the agent threads but is still not used.
>
> Signed-off-by ...

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -181,7 +182,8 @@ static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_
 
-static void son_slave_watchdog(beerocks::config_file::sConfigSlave &beerocks_slave_conf)
+static void son_slave_watchdog(beerocks::config_file::sConfigSlave &beerocks_slave_conf,
+                               const std::unordered_map<int, std::string> &interfaces_map)
```
Can be named "interfaces" (nitpick)

adam1985d replied:
thank you for the comment. it will not be changed in this PR, but I will consider the suggestion in future commits.

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -166,7 +166,7 @@ static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_
         beerocks_slave_conf.enable_repeater_mode[slave_num] == "1";
     son_slave_conf.hostap_iface_type = beerocks::utils::get_iface_type_from_string(
```
No more "before this change - after this change" commits please. Use proper English and describe the change.
Suggested commit message:
> agent: slave_main: use interface name from platform db
>
> Update son slave and slave main to fetch the hostap_iface from the platform DB instead of the configuration file. 
> 
> Signed-off-by ...

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -186,21 +186,25 @@ static void son_slave_watchdog(beerocks::config_file::sConfigSlave &beerocks_sla
                                const std::unordered_map<int, std::string> &interfaces_map)
```
Please remove `beerocks::config_file::sConfigSlave &beerocks_slave_conf` as part of this commit if it is unused.
Same for all functions which used to use the conf file and now do not.

adam1985d replied:
this specific location uses the `beerocks::config_file::sConfigSlave &beerocks_slave_conf` for the `.temp_path` field.
I will change it to pass only the `temp_path` field instead of the whole struct.

adam1985d replied:
other than this location, the rest of the locations still require the `beerocks_slave_conf` as they are using multiple fields from the struct.

tomereli commented on agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:
```diff
@@ -317,19 +317,21 @@ main_thread::main_thread(config_file::sConfigSlave config_,
     auto now = std::chrono::steady_clock::now();
```
Commit message:

>agent: platform_manager: use hostap interface from platform DB
>
>Get the hostap_iface from platform DB instead of the configuration file.
>
>Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d replied:
changed the commit message.

tomereli commented on agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:
```diff
@@ -317,19 +317,21 @@ main_thread::main_thread(config_file::sConfigSlave config_,
     auto now = std::chrono::steady_clock::now();
```
Please remove `config_file::sConfigSlave config_` if its not used.

adam1985d replied:
but it is used

tomereli commented on agent/config/beerocks_agent.conf.in:
```diff
@@ -29,20 +29,17 @@ backhaul_wire_iface=@BEEROCKS_BH_WIRE_IFACE@
 [agent0]
```
Commit message suggestion:

>agent: remove hostap_iface from beerock_agent.conf
>
>The hostap_iface param is now read from the platform DB, thus this
> param is no longer required in the beerocks_agent.conf file and in the
> sConfigSlave struct that is used to read the above-mentioned file.
> Remove them.
>
> Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d replied:
changed the commit message.

tomereli commented on agent/config/beerocks_agent.conf.in:
```diff
@@ -29,20 +29,17 @@ backhaul_wire_iface=@BEEROCKS_BH_WIRE_IFACE@
 [agent0]
 radio_identifier=00:00:00:00:00:00
 hostap_iface_type=WIFI_INTEL
-hostap_iface=@BEEROCKS_WLAN1_IFACE@
```
Also as part of this commit - can't we do the CMAKE changes as well??

adam1985d replied:
the `CMAKE` changes will require `prplmesh_utils.sh.in` changes and also `beerocks_controller.conf.in` changes.

we can create a task for it, but leave it out of this PR for reviewability

tomereli replied:
OK - please add a task for such cleanups.

adam1985d replied:
the following issue will handle the above-mentioned required changes:
https://github.com/prplfoundation/prplMesh/issues/795

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -97,7 +97,7 @@ main() {
         netgear-rax40)
             TARGET=intel_mips
```
Commit message again... Please fix it to align with the CONTRIBUTING.md instructions as I showed in previous commits.

adam1985d replied:
changed the commit message.

tomereli reviewed (changes requested):
Commit messages need heavy refactoring. I practically wrote all the commit messages for you in this review, please try to abide to the commit message formats we normally use in prplmesh which are described in details in the CONTRIBUTING.md.

vitalybu commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -516,7 +516,8 @@ int main(int argc, char *argv[])
     for (size_t i = 0; i < num_of_interfaces; i++) {
         if (beerocks::net::network_utils::linux_iface_exists(interfaces[i].ifname)) {
-            interfaces_map[interfaces[i].radio_num] = interfaces[i].ifname;
+            LOG(DEBUG) << "radio" << i << ".hostap_iface=" << std::string(interfaces[i].ifname);
```
Why do you create an `std::string` just for the logging?
Pretty sure easylogging++ can logs `char[]` just as good.

adam1985d replied:
this change was part of debugging an empty print of hostap_iface read from prplmesh config.
the source of the issue was located elsewhere - but forgot to revert this change.

will change it back to `char[]`

adam1985d committed:
docker: build.sh - update PRPL_FEED hash

Update the PRPL_FEED hash to the latest.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bpl: add cfg_get_hostap_iface API

Preparative commit.
Add infrastructure to get hostap_iface param from the platform DB.
The hostap_iface is available per radio agent (son slave) and if it is
set, a process (beerocks_agent -i<iface>) for that interface is created.
This is instead of the current state which gets the interface name from
the config file beerocks_agent.conf.

Additionally, fix some KW potential issues with the NULL comparison.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bpl: add cfg_get_all_prplmesh_wifi_interfaces API

Preparative commit.
This commit adds infrastructure to get all hostap-interfaces and their
respective radio agents from the prplmesh config file.
This information can be later used to run an agent process per
configured and available interface.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
agent: get hostap_iface from platform DB

Preparative commit.
This commit utilizes the new API to read all the hostap interfaces
from the platform database instead of the agent config file, so that
each platform can have its different interface names.
Interfaces are read into an unordered map of radio index to
interface-name which is propagated to all the agent threads but is
still not used.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
agent: slave_main: use interface name platform DB

Update son slave and slave main to fetch the hostap_iface from the
platform DB instead of the configuration file.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
agent: platform_manager: use hostap interface from platform DB

Get the hostap_iface from platform DB instead of the configuration
file.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
agent: remove hostap_iface from beerock_agent.conf

The hostap_iface param is now read from the platform DB, thus this param
is no longer required in the beerocks_agent.conf file and in the
sConfigSlave struct that is used to read the above-mentioned file.
Remove them.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_async_work_queue.h:
```diff
@@ -20,7 +20,7 @@ namespace beerocks {
 
 class async_work_queue : public thread_base {
 public:
-    async_work_queue()  = default;
```
Looks like a lot of clang-format in this commit. Please at least mention that in the commit message. Is this rebased?

adam1985d committed:
agent: common linux bpl for pc, Turris Omnia and GLInet

Preparative Commit.
The above-mentioned platforms share the same Linux BPL.
The Linux-PC uses wlan0,2 interfaces while the other platforms
use wlan0,1.
BPL returns wlan0-2 as possible interfaces and the upper layer filters
the unsupported interface. This causes a misalignment between
beerocks_agent_conf array of agents to the 2nd interface.
This solution will be removed when the all agent configuration is moved
to BPL.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
Merge branch 'master' into feature/move-hostapd-iface-from-beerocks-config-to-prplmesh-config
